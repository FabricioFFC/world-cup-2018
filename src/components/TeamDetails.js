import React from 'react';
import PropTypes from 'prop-types';
import {
  StyleSheet,
  Image,
  View,
  Text,
  FlatList,
  Dimensions,
} from 'react-native';
import ParallaxScrollView from 'react-native-parallax-scroll-view';

const PARALLAX_HEADER_HEIGHT = 150;
const STICKY_HEADER_HEIGHT = 40;

const window = Dimensions.get('window');
const styles = StyleSheet.create({
  flag: {
    borderWidth: 1,
    borderColor: '#c7c6c6',
    flex: 1,
    height: 22,
    width: 32,
  },
  stickySection: {
    height: STICKY_HEADER_HEIGHT,
  },
  stickySectionText: {
    backgroundColor: '#008558',
    color: 'white',
    fontSize: 20,
    height: STICKY_HEADER_HEIGHT,
    paddingLeft: 10,
    paddingTop: 8,
  },
  badgeContainer: {
    flex: 1,
  },
  badge: {
    alignSelf: 'center',
    height: 150,
    justifyContent: 'center',
    marginTop: 20,
    width: 150,
  },
  statsWrapper: {
    flex: 1,
    flexDirection: 'row',
  },
  statsContainer: {
    alignItems: 'center',
    flex: 1,
    flexDirection: 'column',
    marginLeft: 20,
    marginRight: 20,
    marginTop: 20,
  },
  statDescription: {
    color: 'gray',
    fontSize: 12,
  },
  statValue: {
    color: '#D64545',
    fontSize: 30,
    fontWeight: 'bold',
  },
  playersWrapper: {
    flex: 1,
    flexDirection: 'row',
    marginBottom: 20,
    marginTop: 20,
  },
  playerContainer: {
    alignItems: 'center',
    flex: 1,
    flexDirection: 'column',
    marginTop: 20,
  },
  playerName: {
    color: '#D64545',
    fontSize: 18,
    fontWeight: 'bold',
  },
  playerPosition: {
    color: '#D64545',
    fontSize: 14,
  },
  playerAge: {
    color: '#D64545',
    fontSize: 12,
  },
  playerImage: {
    flex: 1,
    height: 150,
    width: 150,
  },
});

const TeamDetails = ({ team }) => {
  const background = () => (
    <View key="background">
      <Image source={{
        uri: team.profile_image,
        width: window.width,
        height: PARALLAX_HEADER_HEIGHT,
      }}
      />
      <View style={{
        position: 'absolute',
        top: 0,
        width: window.width,
        height: PARALLAX_HEADER_HEIGHT,
      }}
      />
    </View>
  );

  const sticky = () => (
    <View key="sticky-header" style={styles.stickySection}>
      <Text style={styles.stickySectionText}>{team.name}</Text>
    </View>
  );

  const renderPlayer = player => (
    <View style={styles.playerContainer}>
      <Image
        style={styles.playerImage}
        source={{ uri: player.img }}
      />
      <Text style={styles.playerName}>{player.name}</Text>
      <Text style={styles.playerPosition}>{player.position}</Text>
      <Text style={styles.playerAge}>{player.age}</Text>
    </View>
  );

  return (
    <View style={{ flex: 1, backgroundColor: 'white' }}>
      <ParallaxScrollView
        backgroundColor="black"
        parallaxHeaderHeight={PARALLAX_HEADER_HEIGHT}
        stickyHeaderHeight={STICKY_HEADER_HEIGHT}
        renderBackground={background}
        renderStickyHeader={sticky}
      >
        <View style={styles.badgeContainer}>
          <Image
            style={styles.badge}
            source={{ uri: team.badge }}
            resizeMode="contain"
          />
        </View>
        <View style={styles.statsWrapper}>
          <View style={styles.statsContainer}>
            <Text style={styles.statValue}>{team.stats.rank}</Text>
            <Text style={styles.statDescription}>Ranking na FIFA</Text>
          </View>
          <View style={styles.statsContainer}>
            <Text style={styles.statValue}>{team.stats.appearances}</Text>
            <Text style={styles.statDescription}>Presenças na Copa</Text>
          </View>
        </View>
        <View style={styles.statsWrapper}>
          <View style={styles.statsContainer}>
            <Text style={styles.statValue}>{team.stats.titles}</Text>
            <Text style={styles.statDescription}>Títulos</Text>
          </View>
          <View style={styles.statsContainer}>
            <Text style={styles.statValue}>{team.stats.semifinals}</Text>
            <Text style={styles.statDescription}>Semifinais</Text>
          </View>
          <View style={styles.statsContainer}>
            <Text style={styles.statValue}>{team.stats.final}</Text>
            <Text style={styles.statDescription}>Finais</Text>
          </View>
        </View>
        <View style={styles.playersWrapper}>
          <FlatList
            data={team.players}
            numColumns={2}
            keyExtractor={item => item.name}
            renderItem={({ item }) => renderPlayer(item)}
          />
        </View>
      </ParallaxScrollView>
    </View>
  );
};

TeamDetails.propTypes = {
  team: PropTypes.shape({
    flag: PropTypes.string,
    name: PropTypes.string,
    code: PropTypes.string,
  }).isRequired,

};

export default TeamDetails;
