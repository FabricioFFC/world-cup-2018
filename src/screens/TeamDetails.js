import React, { Component } from 'react';
import { View, ActivityIndicator } from 'react-native';

import TeamDetails from './../components/TeamDetails';
import teamsApi from '../api/TeamsApi';

class TeamDetailsScreen extends Component {
  constructor(props) {
    super(props);
    this.state = {
      loading: true,
    };
    this.routeParams = props.navigation.state.params; // eslint-disable-line react/prop-types
  }

  async componentWillMount() {
    const team = await teamsApi.get(this.routeParams.teamCode);
    this.setState({ team, loading: false });
  }

  render() {
    if (this.state.loading) {
      return (
        <View style={{ flex: 1, padding: 20 }}>
          <ActivityIndicator size="large" color="#0000ff" />
        </View>
      );
    }

    return (
      <TeamDetails team={this.state.team} />
    );
  }
}

export default TeamDetailsScreen;
