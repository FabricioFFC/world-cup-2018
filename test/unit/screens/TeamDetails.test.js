import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { ActivityIndicator} from 'react-native';
import nock from 'nock';

import TeamDetailsScreen from '../../../src/screens/TeamDetails';
import TeamDetails from '../../../src/components/TeamDetails';
import brazilMock from '../../fixtures/teams/brazil.json';

const asyncFlush = () => new Promise(resolve => setTimeout(resolve, 100));

describe('TeamDetailsScreen', () => {
  describe('render', () => {
    it('show ActivityIndicator component while team details is not loaded yet', () => {
      // given
      const shallowRenderer = new ShallowRenderer();
      const mockNavigation = {
        state: {
          params: { teamCode: 'BRA' },
        },
      };
      const expectedComponent = <ActivityIndicator
        animating={true} color="#0000ff" hidesWhenStopped={true} size="large"
      />;
      // when
      shallowRenderer.render(<TeamDetailsScreen navigation={mockNavigation}/>);
      const result = shallowRenderer.getRenderOutput();
      // then
      expect(result.props.children).toEqual(expectedComponent);
      expect(result).toMatchSnapshot();
    });

    it('show TeamDetails component when team details is loaded', async() => {
      // given
      const teamCode = 'BRA';
      const mockNavigation = {
        state: {
          params: { teamCode },
        },
      };
      nock('http://private-c09d5b-worldcup20181.apiary-mock.com')
        .get(`/teams/${teamCode}`)
        .reply(200, brazilMock);
      const shallowRenderer = new ShallowRenderer();
      // when
      shallowRenderer.render(<TeamDetailsScreen navigation={mockNavigation}/>);
      const instance = shallowRenderer.getMountedInstance();

      instance.componentDidMount();
      await asyncFlush();

      const result = shallowRenderer.getRenderOutput();
      const instanceRenderedType = instance.render().type;
      // then
      expect(instanceRenderedType).toEqual(TeamDetails);
      expect(result).toMatchSnapshot();
    });
  });
});
