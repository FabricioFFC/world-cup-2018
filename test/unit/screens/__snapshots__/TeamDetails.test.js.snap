// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamDetailsScreen render show ActivityIndicator component while team details is not loaded yet 1`] = `
<Component
  style={
    Object {
      "flex": 1,
      "padding": 20,
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#0000ff"
    hidesWhenStopped={true}
    size="large"
  />
</Component>
`;

exports[`TeamDetailsScreen render show TeamDetails component when team details is loaded 1`] = `
<TeamDetails
  team={
    Object {
      "badge": "https://i.imgur.com/ueaaqJj.png",
      "code": "BRA",
      "flag": "http://img.fifa.com/images/flags/4/bra.png",
      "name": "Brazil",
      "players": Array [
        Object {
          "age": 25,
          "img": "https://api.fifa.com/api/v1/picture/players/2018fwc/308370_sq-150_jpg",
          "name": "A. Becker",
          "position": "Goalkeeper",
        },
        Object {
          "age": 31,
          "img": "https://api.fifa.com/api/v1/picture/players/2018fwc/266773_sq-150_jpg",
          "name": "Cassio",
          "position": "Goalkeeper",
        },
        Object {
          "age": 24,
          "img": "https://api.fifa.com/api/v1/picture/players/2018fwc/395427_sq-150_jpg",
          "name": "Ederson",
          "position": "Goalkeeper",
        },
        Object {
          "age": 33,
          "img": "https://api.fifa.com/api/v1/picture/players/2018fwc/289964_sq-150_jpg",
          "name": "T. Silva",
          "position": "DEFENDER",
        },
        Object {
          "age": 33,
          "img": "https://api.fifa.com/api/v1/picture/players/2018fwc/289958_sq-150_jpg",
          "name": "Miranda",
          "position": "DEFENDER",
        },
        Object {
          "age": 32,
          "img": "https://api.fifa.com/api/v1/picture/players/2018fwc/398286_sq-150_jpg",
          "name": "Geromel",
          "position": "DEFENDER",
        },
        Object {
          "age": 32,
          "img": "https://api.fifa.com/api/v1/picture/players/2018fwc/217161_sq-150_jpg",
          "name": "Filipe Luis",
          "position": "DEFENDER",
        },
        Object {
          "age": 30,
          "img": "https://api.fifa.com/api/v1/picture/players/2018fwc/218284_sq-150_jpg",
          "name": "Marcelo",
          "position": "DEFENDER",
        },
        Object {
          "age": 24,
          "img": "https://api.fifa.com/api/v1/picture/players/2018fwc/332946_sq-150_jpg",
          "name": "Marquinhos",
          "position": "DEFENDER",
        },
        Object {
          "age": 26,
          "img": "https://api.fifa.com/api/v1/picture/players/2018fwc/335656_sq-150_jpg",
          "name": "Danilo",
          "position": "DEFENDER",
        },
        Object {
          "age": 29,
          "img": "https://api.fifa.com/api/v1/picture/players/2018fwc/266774_sq-150_jpg",
          "name": "Fagner",
          "position": "DEFENDER",
        },
        Object {
          "age": 26,
          "img": "https://api.fifa.com/api/v1/picture/players/2018fwc/308386_sq-150_jpg",
          "name": "Casemiro",
          "position": "MIDFIELDER",
        },
        Object {
          "age": 30,
          "img": "https://api.fifa.com/api/v1/picture/players/2018fwc/218292_sq-150_jpg",
          "name": "R. Augusto",
          "position": "MIDFIELDER",
        },
        Object {
          "age": 26,
          "img": "https://api.fifa.com/api/v1/picture/players/2018fwc/308366_sq-150_jpg",
          "name": "P. Coutinho",
          "position": "MIDFIELDER",
        },
        Object {
          "age": 29,
          "img": "https://api.fifa.com/api/v1/picture/players/2018fwc/362727_sq-150_jpg",
          "name": "Paulinho",
          "position": "MIDFIELDER",
        },
        Object {
          "age": 33,
          "img": "https://api.fifa.com/api/v1/picture/players/2018fwc/208016_sq-150_jpg",
          "name": "Fernandinho",
          "position": "MIDFIELDER",
        },
        Object {
          "age": 25,
          "img": "https://api.fifa.com/api/v1/picture/players/2018fwc/394018_sq-150_jpg",
          "name": "Fred",
          "position": "MIDFIELDER",
        },
        Object {
          "age": 29,
          "img": "https://api.fifa.com/api/v1/picture/players/2018fwc/218306_sq-150_jpg",
          "name": "Willian",
          "position": "MIDFIELDER",
        },
        Object {
          "age": 27,
          "img": "https://api.fifa.com/api/v1/picture/players/2018fwc/305070_sq-150_jpg",
          "name": "D. Costa",
          "position": "FORWARD",
        },
        Object {
          "age": 21,
          "img": "https://api.fifa.com/api/v1/picture/players/2018fwc/386559_sq-150_jpg",
          "name": "G. Jesus",
          "position": "FORWARD",
        },
        Object {
          "age": 26,
          "img": "https://api.fifa.com/api/v1/picture/players/2018fwc/314197_sq-150_jpg",
          "name": "Neymar JR",
          "position": "FORWARD",
        },
        Object {
          "age": 26,
          "img": "https://api.fifa.com/api/v1/picture/players/2018fwc/344547_sq-150_jpg",
          "name": "Firmino",
          "position": "FORWARD",
        },
        Object {
          "age": 30,
          "img": "https://api.fifa.com/api/v1/picture/players/2018fwc/329039_sq-150_jpg",
          "name": "Taison",
          "position": "FORWARD",
        },
        Object {
          "age": 57,
          "img": "https://api.fifa.com/api/v1/picture/players/2018fwc/404828_sq-150_jpg",
          "name": "Tite",
          "position": "Coach",
        },
      ],
      "profileImage": "https://img.fifa.com/image/upload/t_tc1/apzbapzlyerna0wn0bcs.jpg",
      "stats": Object {
        "appearances": 20,
        "final": 7,
        "rank": 2,
        "semifinals": 11,
        "titles": 5,
      },
    }
  }
/>
`;
